src/conversions/uuid.rs

```diff
@@ -0,0 +1,301 @@
+#![cfg(feature = "uuid")]
+
+//! Conversions to and from [uuid](https://docs.rs/uuid/latest/uuid/)'s [`Uuid`] type.
+//!
+//! This is useful for converting Python's uuid.UUID into and from a native Rust type.
+//!
+//! # Setup
+//!
+//! To use this feature, add to your **`Cargo.toml`**:
+//!
+//! ```toml
+//! [dependencies]
+#![doc = concat!("pyo3 = { version = \"", env!("CARGO_PKG_VERSION"),  "\", features = [\"uuid\"] }")]
+//! uuid = "1.11.0"
+//! ```
+//!
+//! Note that you must use a compatible version of uuid and PyO3.
+//! The required uuid version may vary based on the version of PyO3.
+//!
+//! # Example
+//!
+//! Rust code to create a function that parses a UUID string and returns it as a `Uuid`:
+//!
+//! ```rust
+//! use pyo3::prelude::*;
+//! use pyo3::exceptions::PyValueError;
+//! use uuid::Uuid;
+//!
+//! #[pyfunction]
+//! fn parse_uuid(s: &str) -> PyResult<Uuid> {
+//!     Uuid::parse_str(s).map_err(|e| PyValueError::new_err(e.to_string()))
+//! }
+//!
+//! #[pymodule]
+//! fn my_module(m: &Bound<'_, PyModule>) -> PyResult<()> {
+//!     m.add_function(wrap_pyfunction!(parse_uuid, m)?)?;
+//!     Ok(())
+//! }
+//! ```
+//!
+//! Python code that validates the functionality
+//!
+//!
+//! ```python
+//! from my_module import parse_uuid
+//! import uuid
+//!
+//! py_uuid = uuid.uuid4()
+//! rust_uuid = parse_uuid(str(py_uuid))
+//!
+//! assert py_uuid == rust_uuid
+//! ```
+use uuid::Uuid;
+
+use crate::conversion::IntoPyObject;
+use crate::exceptions::{PyTypeError, PyValueError};
+use crate::instance::Bound;
+use crate::sync::GILOnceCell;
+use crate::types::any::PyAnyMethods;
+use crate::types::bytearray::PyByteArrayMethods;
+use crate::types::{
+    IntoPyDict, PyByteArray, PyBytes, PyBytesMethods, PyInt, PyStringMethods, PyType,
+};
+use crate::{FromPyObject, Py, PyAny, PyErr, PyObject, PyResult, Python};
+#[allow(deprecated)]
+use crate::{IntoPy, ToPyObject};
+
+static UUID_CLS: GILOnceCell<Py<PyType>> = GILOnceCell::new();
+
+fn get_uuid_cls(py: Python<'_>) -> PyResult<&Bound<'_, PyType>> {
+    UUID_CLS.import(py, "uuid", "UUID")
+}
+
+impl FromPyObject<'_> for Uuid {
+    fn extract_bound(obj: &Bound<'_, PyAny>) -> PyResult<Self> {
+        let py = obj.py();
+
+        if let Ok(uuid_cls) = get_uuid_cls(py) {
+            if obj.is_exact_instance(&uuid_cls) {
+                let uuid_int: u128 = obj.getattr("int")?.extract()?;
+                return Ok(Uuid::from_u128(uuid_int.to_le()));
+            }
+        }
+
+        if obj.is_instance_of::<PyBytes>() || obj.is_instance_of::<PyByteArray>() {
+            let bytes = if let Ok(py_bytes) = obj.downcast::<PyBytes>() {
+                py_bytes.as_bytes()
+            } else if let Ok(py_bytearray) = obj.downcast::<PyByteArray>() {
+                &py_bytearray.to_vec()
```

Comment:
I would suggest we might want to stay away from `Bytearray::to_vec` (see #4736). Maybe for the moment just avoid byte arrays? Or maybe we want to go through the buffer protocol? That's not really much better 🤔